core/domain/stats_jobs_one_off.py

```diff
@@ -195,3 +196,248 @@ def reduce(key, stringified_values):
+            stringified_values: list(str). A list of stringified values
+                associated with the given key. An element of stringified_values
+                would be of the form:
+                    {
+                        'exploration_id': str. the id of the exploration.
+                        'exploration_version': number. the version of the
+                            exploration.
+                        'state_name': str. the name of the corresponding state.
+                        'interaction_id': str. the corresponding interaction ID.
+                    }
+
+        Yields:
+            tuple(str). A 1-tuple whose only element is an error message.
+        """
+        first_value_dict = ast.literal_eval(stringified_values[0])
+        exploration_id = first_value_dict['exploration_id']
+        exploration_version = first_value_dict['exploration_version']
+        state_name = first_value_dict['state_name']
+        interaction_ids = set()
+        for stringified_value in stringified_values:
+            value_dict = ast.literal_eval(stringified_value)
+            interaction_ids.add(value_dict['interaction_id'])
+        if len(interaction_ids) != 1:
+            yield (
+                'Expected exploration %s (version %s) and state %s pair to '
+                'only have 1 interaction ID, but found: %s' % (
+                    exploration_id, exploration_version, state_name,
+                    interaction_ids))
+
+
+class PurgeNewAnswersAuditJob(jobs.BaseMapReduceJobManager):
+    """A pre-purge audit job.
+
+    This job performs the exact same work as PurgeNewAnswersJob without the
+    actual deletion. This should be run before the deletion to ensure only the
+    expected answer buckets are removed.
+    """
+    _ANSWER_BUCKET_REMOVED_KEY = 'removed_answer_bucket'
+    _ANSWER_BUCKET_RETAINED_KEY = 'retained_answer_bucket'
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [stats_models.StateAnswersModel]
+
+    @staticmethod
+    def map(item):
+        """Implements the map function and yields all answer buckets associated
+        with interactions that no longer exist.
+
+        Args:
+            item: StateAnswersModel.
+
+        Yields:
+            tuple. a 2-tuple in the form
+                (output_key, value).
+                'output_key': str. either _ANSWER_BUCKET_RETAINED_KEY or
+                    _ANSWER_BUCKET_REMOVED_KEY with an interaction ID appended.
+                'value': a dict, whose structure is as follows:
+                    {
+                        'item_id': str. the id of the mapped model.
+                        'interaction_id': str. the corresponding interaction ID.
+                    }
+        """
+        interaction_ids = (
+            interaction_registry.Registry.get_all_interaction_ids())
+        if item.interaction_id not in interaction_ids:
+            removal_key = '%s-%s' % (
```

Comment:
Optional: Here and below the delimiter doesn't matter, but you might want to consider using ':' always for consistency.